File: 백준/Gold/10986. 나머지 합/나머지 합.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        
        long[] prefixSum = new long[N + 1];
        long[] count = new long[M];
        
        st = new StringTokenizer(br.readLine());
        for (int i = 1; i <= N; i++) {
            int num = Integer.parseInt(st.nextToken());
            prefixSum[i] = (prefixSum[i-1] + num) % M;
            if (prefixSum[i] < 0) {
                prefixSum[i] += M;
            }
            count[(int)prefixSum[i]]++;
        }
        
        long result = 0;
        result += count[0];

        for (int i = 0; i < M; i++) {
            if (count[i] >= 2) {
                result += count[i] * (count[i] - 1) / 2;
            }
        }
        
        System.out.println(result);
    }
}

File: 백준/Gold/10986. 나머지 합/나머지 합.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.StringTokenizer

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val st = StringTokenizer(br.readLine())
    
    val n = st.nextToken().toInt()
    val m = st.nextToken().toInt()
    
    val prefixSum = LongArray(n + 1)
    val count = LongArray(m)
    
    val nums = StringTokenizer(br.readLine())
    for (i in 1..n) {
        val num = nums.nextToken().toInt()
        prefixSum[i] = (prefixSum[i - 1] + num) % m
        if (prefixSum[i] < 0) {
            prefixSum[i] = prefixSum[i] + m
        }
        count[prefixSum[i].toInt()]++
    }
    
    var result = 0L
    result += count[0]

    for (i in 0 until m) {
        if (count[i] >= 2) {
            result += count[i] * (count[i] - 1) / 2
        }
    }
    
    println(result)
}